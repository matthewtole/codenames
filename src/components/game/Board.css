.BoardWrapper {
  flex: 1 1;
}

.Board {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  padding: 5px;
}

.BoardRow {
  display: flex;
  flex-direction: row;
  height: 20%;
}

@media only screen and (max-width: 1024px) {
  .Board {
    padding: 3px;
  }
}
