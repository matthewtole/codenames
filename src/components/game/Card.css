.Card {
  width: 20%;
  font-size: 1.75rem;
  /* margin: 1px; */
  user-select: none;
  position: relative;
}

.Card > div {
  background-color: #eee;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  position: absolute;
  left: 0;
  right: 0;
  bottom: 0;
  top: 0;
  border-radius: 4px;
  transition: margin 0.1s ease-out;
  margin: 0;
  border: 2px solid #fff;
}

.Card--revealed {
  text-decoration: line-through;
  opacity: 0.2;
}

.Card--controller[data-role='BLUE_SPY'] > div {
  background-color: #bbdefb;
}

.Card--controller[data-role='RED_SPY'] > div {
  background-color: #ef9a9a;
}

.Card--controller[data-role='BYSTANDER'] > div {
  background-color: #ff9100;
}

.Card--controller[data-role='ASSASSIN'] > div {
  background-color: #999;
}

.Card[data-role='BLUE_SPY'].Card--revealed > div {
  background-color: #2196f3;
  color: #bbdefb;
}

.Card[data-role='RED_SPY'].Card--revealed > div {
  background-color: #d32f2f;
  color: #ef9a9a;
}

.Card[data-role='BYSTANDER'].Card--revealed > div {
  background-color: #ff9100;
}

.Card[data-role='ASSASSIN'].Card--revealed > div {
  background-color: #222;
}

.Card--highlighted > div {
  /* border: 5px solid #000; */
  /* box-shadow: 0 0 20px rgba(0, 0, 0, 0.5); */
  box-shadow: 0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22);
  margin: -7px;
  z-index: 200;
}

@media only screen and (max-width: 1024px) {
  .Card {
    font-size: 1.2rem;
  }
}
