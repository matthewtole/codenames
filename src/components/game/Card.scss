@import '../../sass/_variables';

.Card {
  width: 20%;
  font-size: 1.75rem;
  user-select: none;
  position: relative;

  & > div {
    background-color: #eee;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
    position: absolute;
    left: 0;
    right: 0;
    bottom: 0;
    top: 0;
    border-radius: 4px;
    margin: 5px;
    border: 2px solid #ddd;

    box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
    transition: all 0.3s cubic-bezier(0.25, 0.8, 0.25, 1);
  }

  &--revealed > div {
    border: 0;
  }

  &--highlighted > div {
    box-shadow: 0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22);
    // margin: -7px;
    z-index: 10;
  }

  &--revealed {
    text-decoration: line-through;
    opacity: 0.2;
  }

  &[data-role='BLUE_SPY'] {
    &.Card--controller,
    &.Card--game-over {
      > div {
        background-color: $blue-spy--light;
        border-color: $blue-spy--dark;
      }

      &.Card--controller.Card--highlighted > div {
        background-color: $blue-spy--dark;
        color: #fff;
      }
    }

    &.Card--revealed {
      > div {
        background-color: $blue-spy--dark;
        color: $blue-spy--light;
      }
    }
  }

  &[data-role='RED_SPY'] {
    &.Card--controller,
    &.Card--game-over {
      > div {
        background-color: $red-spy--light;
        border-color: $red-spy--dark;
      }
    }

    &.Card--controller.Card--highlighted > div {
      background-color: $red-spy--dark;
      color: #fff;
    }

    &.Card--revealed {
      > div {
        background-color: $red-spy--dark;
        color: $red-spy--light;
      }
    }
  }

  &[data-role='BYSTANDER'] {
    &.Card--controller,
    &.Card--game-over {
      > div {
        background-color: $bystander--light;
        border-color: $bystander--dark;
      }
    }

    &.Card--controller.Card--highlighted > div {
      background-color: $bystander--dark;
      color: #fff;
    }

    &.Card--revealed {
      > div {
        background-color: $bystander--dark;
      }
    }
  }

  &[data-role='ASSASSIN'] {
    &.Card--controller,
    &.Card--game-over {
      > div {
        background-color: $assassin--light;
        border-color: $assassin--dark;
      }
    }

    &.Card--controller.Card--highlighted > div {
      background-color: $assassin--dark;
      color: #fff;
    }

    &.Card--revealed {
      > div {
        background-color: $assassin--dark;
      }
    }
  }
}

@media only screen and (max-width: 1024px) {
  .Card {
    font-size: 1.2rem;

    & > div {
      margin: 3px;
    }
  }
}
