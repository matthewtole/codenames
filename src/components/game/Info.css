.Info {
  display: flex;
  flex-direction: row;
  height: 48px;
  justify-content: space-between;
  flex-shrink: 0;
  flex-grow: 0;
}

.Info--winner {
  height: 144px;
}

.Info--red {
  background-color: #ef9a9a;
}

.Info--blue {
  background-color: #bbdefb;
}

.Info__turn {
  flex: 1;
  display: flex;
  align-items: center;
  font-size: 24px;
  line-height: 24px;
  text-transform: uppercase;
}

.Info--winner .Info__turn {
  font-size: 48px;
  line-height: 48px;
  font-weight: 100;
}

.Info__turn .fa {
  margin: 0 4px;
}

.Info--red .Info__turn {
  justify-content: flex-end;
}

.Info--blue .Info__turn {
  justify-content: flex-start;
}

.Info__turn.Info__turn--winner {
  justify-content: center;
}

.Info__count {
  flex: 0 1;
  display: flex;
  flex-direction: row;
  margin: 0 4px;
}

.Info__counter {
  height: 32px;
  width: 32px;
  margin: 8px 4px;
  border: 2px solid #fff;
  border-radius: 8px;
  display: flex;
  align-items: center;
  justify-content: center;
}

.Info__counter i.fa {
  color: rgba(255, 255, 255, 0.8);
}

.Info__count--blue .Info__counter {
  background-color: #2196f3;
}

.Info__count--red .Info__counter {
  background-color: #d32f2f;
}

@media only screen and (max-width: 1024px) {
  .Info__counter {
    width: 28px;
    margin: 8px 2px;
  }

  .Info__counter i.fa {
    font-size: 12px;
  }
}
