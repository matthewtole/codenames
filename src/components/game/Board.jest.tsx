import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { BoardProps, BoardMode, Board } from './Board';
import { OriginalCards1 } from '../../fixtures/cards';

describe('Board', () => {
  const defaultProps: BoardProps = {
    cards: OriginalCards1,
    width: 5,
    height: 5,
    revealedCards: [],
    mode: BoardMode.CONTROLLER,
    isGameOver: false,
    onRevealCard: jest.fn(),
    onHighlightCard: jest.fn(),
  };

  it('renders correctly', () => {
    const tree = renderer.create(<Board {...defaultProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
