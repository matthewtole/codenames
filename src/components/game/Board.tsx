import * as React from 'react';
import './Board.css';
import { Card as CardComponent } from './Card';
import {
  Card,
  Coordinate,
  Role,
  CoordinateValue,
} from '../../redux/game/types';

export enum BoardMode {
  VIEWER = 'VIEWER',
  CONTROLLER = 'CONTROLLER',
}

export interface BoardProps {
  width: number;
  height: number;
  cards: Card[];
  revealedCards: number[];
  mode: BoardMode;
  highlighted?: Coordinate;
  highlightedRow?: number;
  isGameOver: boolean;
  onRevealCard: (card: Coordinate) => void;
  onHighlightCard: (card?: Coordinate) => void;
}

export class Board extends React.PureComponent<BoardProps, {}> {
  constructor(props: BoardProps) {
    super(props);
    this.state = {};
  }

  render() {
    return (
      <div className="BoardWrapper" onClick={() => this.highlightCard()}>
        <div className="Board">
          {this.renderRowHighlight()}
          {Array(this.props.height)
            .fill(0)
            .map((_: number, row: CoordinateValue) => this.renderRow(row))}
        </div>
      </div>
    );
  }

  private renderRowHighlight() {
    if (this.props.highlightedRow === undefined) {
      return null;
    }
    const row = this.props.highlightedRow;
    return (
      <div className="Board__row-highlight-wrapper">
        <div
          className="Board__row-highlight"
          style={{ top: 0, height: `${row * 20}%` }}
        />
        <div
          className="Board__row-highlight"
          style={{
            top: `${(row + 1) * 20}%`,
            height: `${(5 - row + 1) * 20}%`,
          }}
        />
      </div>
    );
  }

  private renderRow = (row: CoordinateValue) => {
    return (
      <div className="BoardRow" key={`row_${row}`}>
        {Array(this.props.width)
          .fill(0)
          .map((__: number, col: CoordinateValue) => this.renderCard(row, col))}
      </div>
    );
  }

  private renderCard = (row: CoordinateValue, col: CoordinateValue) => {
    const index = this.coordinateToIndex({ row, col });
    return (
      <CardComponent
        key={`card_${row}_${col}`}
        word={this.getWord(index)}
        role={this.getRole(index)}
        revealed={this.isRevealed(index)}
        highlighted={this.isHighlighted({ row, col })}
        onHighlight={() => {
          this.highlightCard({ row, col });
        }}
        onReveal={() => {
          this.revealCard({ row, col });
        }}
        boardMode={this.props.mode}
        isGameOver={this.props.isGameOver}
      />
    );
  }

  private getWord = (index: number): string => {
    return this.props.cards[index].word;
  }

  private getRole = (index: number): Role => {
    return this.props.cards[index].role;
  }

  private isRevealed = (index: number): boolean => {
    return this.props.revealedCards.indexOf(index) >= 0;
  }

  private isHighlighted = (card: Coordinate): boolean => {
    if (this.props.highlighted === undefined) {
      return false;
    }
    return (
      this.props.highlighted.row === card.row &&
      this.props.highlighted.col === card.col
    );
  }

  private highlightCard = (card?: Coordinate) => {
    this.props.onHighlightCard(card);
  }

  private revealCard = (card: Coordinate) => {
    this.props.onRevealCard(card);
  }

  private coordinateToIndex(coordinate: Coordinate): number {
    return coordinate.row * this.props.width + coordinate.col;
  }
}
