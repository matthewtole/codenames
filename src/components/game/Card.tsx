import * as React from 'react';
import './Card.css';
import { Role } from '../../lib/types';
import * as cx from 'classnames';
import { BoardMode } from './Board';

export interface CardProps {
  word: string;
  role: Role;
  revealed: boolean;
  highlighted: boolean;
  boardMode: BoardMode;
  onReveal: () => void;
  onHighlight: () => void;
}

export class Card extends React.PureComponent<CardProps, object> {
  render() {
    return (
      <div
        className={cx({
          Card: true,
          'Card--highlighted': this.props.highlighted,
          'Card--revealed': this.props.revealed,
          [`Card--${this.props.boardMode}`]: true,
        })}
        data-role={this.props.role}
        onClick={this.handleClick}
      >
        <div>{this.props.word}</div>
      </div>
    );
  }

  handleClick = (e: React.MouseEvent<HTMLDivElement>) => {
    e.stopPropagation();
    if (this.props.revealed) {
      return;
    }
    if (this.props.highlighted) {
      this.props.onReveal();
    } else {
      this.props.onHighlight();
    }
  }
}
