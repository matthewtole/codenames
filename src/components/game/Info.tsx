import * as React from 'react';
import { Team } from '../../redux/game/types';
import './Info.scss';
import * as cx from 'classnames';

export interface InfoProps {
  turn: Team;
  winner?: Team;
  spyCounts: { [key: string]: number };
  showMenu: boolean;
  onMenuOpen: () => void;
}

interface SpyCounterProps {
  team: Team;
  count: number;
  winner?: Team;
}

class SpyCounter extends React.Component<SpyCounterProps, {}> {
  render() {
    const icons = [];
    for (let c = 0; c < this.props.count; c += 1) {
      icons.push(
        <div className="Info__counter" key={c}>
          <i className="fa fa-user-secret fa-fw" />
        </div>
      );
    }
    return (
      <div
        className={cx({
          Info__count: true,
          'Info__count--blue': this.props.team === Team.BLUE,
          'Info__count--red': this.props.team === Team.RED,
        })}
      >
        {icons}
      </div>
    );
  }
}

export class Info extends React.Component<InfoProps, {}> {
  render() {
    const { turn, winner, spyCounts, showMenu } = this.props;
    let contents = null;
    if (winner !== undefined) {
      contents = (
        <div className="Info__turn Info__turn--winner">
          THE {winner} TEAM WINS
        </div>
      );
    } else {
      contents = [
        <SpyCounter
          key="spy-counter-blue"
          team={Team.BLUE}
          count={spyCounts[Team.BLUE]}
          winner={winner}
        />,
        <div className="Info__turn" key="turn">
          {turn === Team.RED ? null : <i className="fa fa-angle-double-left" />}
          {turn}
          {turn === Team.BLUE ? null : (
            <i className="fa fa-angle-double-right fa-fw" />
          )}
        </div>,
        <SpyCounter
          key="spy-counter-red"
          team={Team.RED}
          count={spyCounts[Team.RED]}
          winner={winner}
        />,
      ];
    }
    return (
      <div
        className={cx({
          Info: true,
          'Info--red': turn === Team.RED,
          'Info--blue': turn === Team.BLUE,
          'Info--winner': !!winner,
        })}
      >
        {contents}
        {showMenu && this.renderMenu()}
      </div>
    );
  }

  private renderMenu = () => {
    return (
      <button
        className="GameControls_menu"
        onClick={() => this.props.onMenuOpen()}
        key="menu"
      >
        <i className="fas fa-bars fa-fw" />
      </button>
    );
  }
}
