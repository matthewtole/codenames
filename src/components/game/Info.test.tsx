import * as React from 'react';
import { Info, InfoProps } from './Info';
import * as renderer from 'react-test-renderer';
import { Team } from '../../redux/game/types';

describe('Info', () => {
  const defaultProps: InfoProps = {
    turn: Team.RED,
    spyCounts: {
      [Team.RED]: 5,
      [Team.BLUE]: 3,
    },
    onMenuOpen: jest.fn(),
    onClearHighlight: jest.fn(),
    showMenu: true,
  };

  it('renders correctly', () => {
    const info = renderer.create(<Info {...defaultProps} />).toJSON();
    expect(info).toMatchSnapshot();
  });
});
