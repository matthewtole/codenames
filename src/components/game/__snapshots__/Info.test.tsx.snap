// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Info renders correctly 1`] = `
<div
  className="Info Info--red"
>
  <div
    className="Info__count Info__count--blue"
  >
    <div
      className="Info__counter"
    />
    <div
      className="Info__counter"
    />
    <div
      className="Info__counter"
    />
  </div>
  <div
    className="Info__turn"
  >
    RED
    <i
      className="fa fa-angle-double-right fa-fw"
    />
  </div>
  <div
    className="Info__count Info__count--red"
  >
    <div
      className="Info__counter"
    />
    <div
      className="Info__counter"
    />
    <div
      className="Info__counter"
    />
    <div
      className="Info__counter"
    />
    <div
      className="Info__counter"
    />
  </div>
  <button
    className="GameControls_menu"
    onClick={[Function]}
  >
    <i
      className="fas fa-bars fa-fw"
    />
  </button>
</div>
`;
